Mathlib/Topology/Algebra/Category/ProfiniteGrp/Limits.lean

```diff
@@ -0,0 +1,147 @@
+  sending an open normal subgroup `U` to `P ⧸ U`, where `P : ProfiniteGrp`.
+
+* `toLimit` : The continuous homomorphism from a profinite group `P` to
+  the projective limit of its quotients by open normal subgroups ordered by inclusion.
+
+* `ContinuousMulEquivLimittoFiniteQuotientFunctor` : The `toLimit` is a
+  `ContinuousMulEquiv`
+
+## Main Statements
+
+* `OpenNormalSubgroupSubClopenNhdsOfOne` : For any open neighborhood of `1` there is an
+  open normal subgroup contained in it.
+
+-/
+
+universe u
+
+open CategoryTheory TopologicalGroup
+
+namespace ProfiniteGrp
+
+instance (P : ProfiniteGrp) : SmallCategory (OpenNormalSubgroup P) :=
+  Preorder.smallCategory (OpenNormalSubgroup ↑P.toProfinite.toTop)
+
+/-- The functor from `OpenNormalSubgroup P` to `FiniteGrp` sending `U` to `P ⧸ U`,
+where `P : ProfiniteGrp`. -/
+def toFiniteQuotientFunctor (P : ProfiniteGrp) : OpenNormalSubgroup P ⥤ FiniteGrp := {
+    obj := fun H => FiniteGrp.of (P ⧸ H.toSubgroup)
+    map := fun fHK => QuotientGroup.map _ _ (.id _) (leOfHom fHK)
+    map_id _ := QuotientGroup.map_id _
+    map_comp f g := (QuotientGroup.map_comp_map
+      _ _ _ (.id _) (.id _) (leOfHom f) (leOfHom g)).symm }
+
+/--The `MonoidHom` from a profinite group `P` to  the projective limit of its quotients by
+open normal subgroups ordered by inclusion.-/
+def toLimit_fun (P : ProfiniteGrp.{u}) : P →*
+    limit (toFiniteQuotientFunctor P ⋙ forget₂ FiniteGrp ProfiniteGrp) where
+  toFun p := ⟨fun _ => QuotientGroup.mk p, fun _ => rfl⟩
+  map_one' := Subtype.val_inj.mp rfl
+  map_mul' _ _ := Subtype.val_inj.mp rfl
+
+lemma toLimit_fun_continuous (P : ProfiniteGrp.{u}) : Continuous (toLimit_fun P) := by
+  apply continuous_induced_rng.mpr (continuous_pi _)
+  intro H
+  dsimp only [Functor.comp_obj, CompHausLike.toCompHausLike_obj, CompHausLike.compHausLikeToTop_obj,
+    CompHausLike.coe_of, Functor.comp_map, CompHausLike.toCompHausLike_map,
+    CompHausLike.compHausLikeToTop_map, Set.mem_setOf_eq, toLimit_fun,
+    MonoidHom.coe_mk, OneHom.coe_mk, Function.comp_apply]
+  apply Continuous.mk
+  intro s _
+  rw [← (Set.biUnion_preimage_singleton QuotientGroup.mk s)]
+  refine isOpen_iUnion (fun i ↦ isOpen_iUnion (fun _ ↦ ?_))
+  convert IsOpen.leftCoset H.toOpenSubgroup.isOpen' (Quotient.out i)
+  ext x
+  simp only [Set.mem_preimage, Set.mem_singleton_iff]
+  nth_rw 1 [← QuotientGroup.out_eq' i, eq_comm, QuotientGroup.eq]
+  exact Iff.symm (Set.mem_smul_set_iff_inv_smul_mem)
+
+/-- The morphism in the category of `ProfiniteGrp` from a profinite group `P` to
+the projective limit of its quotients by open normal subgroups ordered by inclusion.-/
+def toLimit (P : ProfiniteGrp.{u}) : P ⟶
+    limit (toFiniteQuotientFunctor P ⋙ forget₂ FiniteGrp ProfiniteGrp) := {
+  toLimit_fun P with
+  continuous_toFun := toLimit_fun_continuous P }
+
+theorem denseRange_toLimit (P : ProfiniteGrp.{u}) : DenseRange (toLimit P) := by
```

Comment:
Maybe add a comment above this theorem that it is just an auxiliary result needed to prove `toLimit_surjective`?